Components: Expand theming support in COLORS

mirka opened the pull request:
Part of #44116

## What?

Improves support for component theming in the `COLORS` object.

## Why?

To move toward wider support for component theming via the `Theme` component.

## Testing Instructions

This shouldn't cause any visual changes, except when using `<Theme>` with any components that use variables from `COLORS.ui`.

mirka committed:
Components: Expand theming support in COLORS

mirka committed:
Update changelog

mirka commented on packages/components/src/utils/colors-values.js:
```diff
@@ -36,34 +36,57 @@ const ADMIN = {
+	borderFocus: THEME.accent,
+	borderDisabled: THEME.gray[ 400 ],
+	textDisabled: THEME.gray[ 600 ],
+	textDark: white, // TODO: Make this Theme-ready
```
The one component that uses this `textDark` isn't working correctly right now, so we'll address this later.

ciampo replied:
Could we use the `--wp-components-color-foreground-inverted` variable here?

mirka replied:
Yes. You know, I didn't want to touch any components in this PR (f0f2710), but I'll just do this one now. I also brought in all the vars we currently have in theme-variables.scss so we can work from a clean, synced palette (e6027506ab437dcb6c3659dbf57b6fc2273c2e86).

mirka commented on packages/components/src/utils/colors-values.js:
```diff
@@ -36,34 +36,57 @@ const ADMIN = {
 	/**
 	 * The main gray color object.
+	 *
+	 * @deprecated Use semantic aliases in `COLORS.ui` or theme-ready variables in `COLORS.theme.gray`.
```
This deprecation marking should encourage devs to stop using this while we migrate the existing usages.

mirka committed:
Update snapshots

ciampo commented on packages/components/src/utils/colors-values.js:
```diff
@@ -36,34 +36,57 @@ const ADMIN = {
+	textDark: white, // TODO: Make this Theme-ready
+
+	// Matches @wordpress/base-styles
+	darkGrayPlaceholder: rgba( GRAY[ 900 ], 0.62 ), // TODO: Make this work with CSS vars
+	lightGrayPlaceholder: rgba( white, 0.65 ), // TODO: Make this work with CSS vars
```
Could we use the [`color-mix` CSS function](https://developer.mozilla.org/en-US/docs/Web/CSS/color_value/color-mix) like we did in the progress bar? 

https://github.com/WordPress/gutenberg/blob/05693fb1b9c1128a946345fa936cf8588067e842/packages/components/src/progress-bar/styles.ts#L30-L35

mirka replied:
Oh, good one. Implemented in c9020bd.

ciampo commented on packages/components/src/utils/colors-values.js:
```diff
@@ -36,34 +36,57 @@ const ADMIN = {
+	background: `var(--wp-components-color-background, ${ white })`,
+	foreground: `var(--wp-components-color-foreground, ${ GRAY[ 900 ] })`,
+	gray: {
+		900: `var(--wp-components-color-gray-900, ${ GRAY[ 900 ] })`,
```
It doesn't look like `--wp-components-color-gray-900` is defined in the `Theme` component. I can't remember if there a specific reason behind that decision, or if we just forgot to add it.

mirka replied:
Good catch! The reason was that `gray-900` is the same as `foreground` and thus redundant. Addressed in 8b321ab by keeping a deprecated alias that nudges people to use `foreground`.

Ultimately I feel like the scale should be named like `foreground-600` etc. instead of `gray-600`, but I'm not quite ready to commit until we do a bit more actual migrations and exploration.

ciampo commented on packages/components/src/utils/colors-values.js:
```diff
@@ -36,34 +36,57 @@ const ADMIN = {
+	borderHover: THEME.gray[ 700 ],
+	borderFocus: THEME.accent,
+	borderDisabled: THEME.gray[ 400 ],
+	textDisabled: THEME.gray[ 600 ],
```
Should we add a `text` alias under UI ? Currently components would be forced to use `theme.foreground` instead.

mirka replied:
I'm actually kind of reluctant to add a ton of semantic aliases for the pretty obvious ones like `foreground`, `background`, and `accent` because it adds more overhead when reworking and renaming stuff. I do think the aliases for the grayscales are useful because nobody memorizes them and it's easy to introduce inconsistencies. Does that resonate?

ciampo replied:
My original reasoning behind proposing `text` was mostly so that folks can continue to use "semantic" color aliases to style their components, instead of mixing them with "raw" theme values. But if we're ok with folks using both `COLORS.ui.*` and `COLORS.theme.*`, then we can keep things as-is and only adjust later / if needed, especially if the long term plan is to move theming functionality to a separate package.

ciampo reviewed (commented):
It's exciting to see some CSS-in-JS-based components starting to follow theme variables!

mirka committed:
Fix gray-900 var

I'll leave this here with a deprecation marking, just to inform people that they should be using `foreground` when they try and reach for gray-900.

mirka committed:
Completely match theme-variables scss

mirka committed:
Replace `ui.textDark` to `theme.foregroundInverted`

mirka committed:
Replace with color-mix

ciampo commented on packages/components/src/utils/colors-values.js:
```diff
@@ -29,41 +27,65 @@ const ALERT = {
+	 * @deprecated Use semantic aliases in `COLORS.ui` or theme-ready variables in `COLORS.theme.gray`.
 	 */
-	gray: GRAY,
+	gray: GRAY, // TODO: Stop exporting this when everything is migrated to `theme` or `ui`
```
Do we have a tracking issue for all the planned work that we intend to do around theming in the short term?


Edit: yes, we do! It's https://github.com/WordPress/gutenberg/issues/44116.  We could probably tidy that issue up, knowing that in the long term the `Theme` component in `@wordpress/components` will be removed and replaced by functionality from `@wordpress/theme`.

ciampo reviewed (approved):
🚀

mirka committed:
Merge branch 'trunk' into add/theme-colors-js

mirka committed:
Fix changelog placement
